File: src/ctci/java/Ch1/Q3.java
package Ch1;

import java.util.Scanner;

/**
 * Created by Michael on 3/15/2017.
 *
 * Decide if one string is a permutation of the other
 */
public class Q3 {

    public static void main(String[] args) {
        System.out.println("Decide if one string is a permutation of the other");
        Scanner in = new Scanner(System.in);
        while (true) {
            System.out.print("Your inputs, separated by whitespace: ");
            String input1 = in.next();
            String input2 = in.next();
            boolean result1 = isPermutation(input1, input2);
            System.out.println(input1 + " and " + input2 + " are a permutation? (T/F): " + result1);
            System.out.println();

        }
    }

    public static boolean isPermutation(String str1, String str2) {
        int rangeOfCharacters = 128; // assume ascii
        int[] counter = new int[rangeOfCharacters]; // number of characters we are using

        if (str1.length() != str2.length()) {
            return false;
        }

        for (int i = 0; i < str1.length(); i++) {
            int curr = str1.charAt(i);
            counter[curr]++;
        }

        for (int j = 0; j < str2.length(); j++) {
            int curr = str2.charAt(j);
            counter[curr]--;
            if (counter[curr] < 0) {
                return false;
            }
        }
        return true;
    }


}

File: src/ctci/python/Ch1/Q3.py
# Decide if one string is a permutation of the other

def isPermutation(str1, str2):
    ascii = 128 # assume ascii, or use the size of possible chars in whatever set you're interested in
    counter = [0 for i in range(0, ascii)]
    if len(str1) != len(str2):
        return False
    for c in str1:
        index = ord(c)
        counter[index] += 1
    for c in str2:
        index = ord(c)
        counter[index] -= 1
        if counter[index] < 0:
            return False
    return True


if __name__ == '__main__':
    print "Decide if one string is a permutation of the other"
    while True:
        print "Input string: "
        strIn = raw_input().split()
        str1 = strIn[0]
        str2 = strIn[1]
        print "received: ", strIn
        result1 = isPermutation(str1, str2)
        print "Result: ", result1
        print